ggshield/oauth.py

```diff
@@ -0,0 +1,273 @@
+        )
+        webbrowser.open_new_tab(request_uri)
+
+    def _prepare_server(self) -> None:
+        try:
+            self.server = HTTPServer(
+                ("127.0.0.1", REDIRECT_PORT), self._handler_wrapper.request_handler
+            )
+        except OSError:
+            raise click.ClickException(f"Port {REDIRECT_PORT} is already in use.")
+
+    def _wait_for_code(self) -> None:
+        """
+        Open a local server on the predetermined port to receive the post-login redirection
+        including the authorazation code.
+        It uses a special request handler which is able to kill the server once the code
+        is received
+        return the authorization code or None if none could be fetched.
+        """
+        try:
+            while not self._handler_wrapper.complete:
+                self.server.handle_request()  # type: ignore
+        except KeyboardInterrupt:
+            raise click.ClickException("Aborting")
+
+        if self._handler_wrapper.error_message is not None:
+            raise click.ClickException(self._handler_wrapper.error_message)
+
+    def _get_code(self, uri: str) -> str:
+        """
+        extract the authorization from the incoming request uri
+        if no code can be extracted, return None
+        """
+        try:
+            authorization_code = self._oauth_client.parse_request_uri_response(uri).get(
+                "code"
+            )
+        except OAuth2Error:
+            authorization_code = None
+        if authorization_code is None:
+            raise OAuthError("Invalid code received from the callback.")
+        return authorization_code  # type: ignore
+
+    def _claim_token(self, authorization_code: str) -> None:
+        """
+        Exchange the authorization code with a valid access token using GitGuardian public api.
+        If no valid token could be retrieved, exit the authentication process with an error message
+        """
+
+        request_params = {"name": self._token_name}
+
+        request_body = self._oauth_client.prepare_request_body(
+            code=authorization_code,
+            redirect_uri=self.redirect_uri,
+            code_verifier=self.code_verifier,
+            body=urlparse.urlencode(request_params),
+        )
+
+        response = requests.post(
+            urlparse.urljoin(self.api_url, "oauth/token"),
+            request_body,
+            headers={"Content-Type": "application/x-www-form-urlencoded"},
+        )
+
+        if not response.ok:
+            raise OAuthError("Cannot retrieve token.")
+
+        self._access_token = response.json()["key"]
+        self.config.auth_config.current_token = self._access_token
+
+    def _validate_access_token(self) -> Dict[str, Any]:
+        """
+        Validate the token using GitGuardian public api.
+        If the token is not valid, exit the authentication process with an error message.
+        """
+        client = retrieve_client(self.config)
```

Comment:
:art: the client keeps a session, it would be better to reuse the same client everywhere in this class.

a `client` cached_property should do.